kafka-client/src/main/java/dev/responsive/kafka/clients/ResponsiveRestoreConsumer.java

```diff
@@ -1,61 +1,111 @@
  * minimizing the length of restores when using a Responsive Store over a compacted changelog
  * topic (which cannot be truncated).
+ * <p>
+ * This consumer is used by Kafka Streams for both active and standby task restoration, though
+ * only one or the other at a given time. This may be subject to change, but the general process
+ * follows this pattern:
+ * 1. initialize changelogs for newly assigned tasks:
+ *     i. restoreConsumer.assign(current assignment + new changelogs)
+ *     ii. if STANDBY_MODE: restoreConsumer.pause(standby changelogs)
+ *     iii. restoreConsumer.seekToBeginning(new changelogs)
+ *     iv. startOffset = restoreConsumer.position(changelog)
+ *     iv. for each ACTIVE changelog: call StateRestoreListener#onRestoreStart
+ * 2. restore active/standby state stores:
+ *     i. poll for new record batch
+ *     ii. restore batch through StateRestoreCallback
+ *     iii. for each ACTIVE changelog in batch: call StateRestoreListener#onBatchRestored
+ *     iv. update current offset based on highest offset in batch
  *
  * @param <K> Type of the record keys
  * @param <V> Type of the record values
  */
 public class ResponsiveRestoreConsumer<K, V> extends DelegatingConsumer<K, V> {
-  final Function<TopicPartition, OptionalLong> startOffsets;
+  private final Logger log;
+
+  private final Function<TopicPartition, OptionalLong> startOffsets;
+  private final Set<TopicPartition> uninitializedOffsets = new HashSet<>();
 
   public ResponsiveRestoreConsumer(
+      final String clientId,
       final Consumer<K, V> delegate,
       final Function<TopicPartition, OptionalLong> startOffsets
   ) {
     super(delegate);
     this.startOffsets = Objects.requireNonNull(startOffsets);
+    this.log = new LogContext(
+        String.format("responsive-restore-consumer [%s]", Objects.requireNonNull(clientId))
+    ).logger(ResponsiveConsumer.class);
   }
 
-  private List<TopicPartition> seekInitial(final Collection<TopicPartition> partitions) {
-    final List<TopicPartition> notFound = new LinkedList<>();
+  private Set<TopicPartition> initializeOffsets(final Collection<TopicPartition> partitions) {
+    final Set<TopicPartition> notFound = new LinkedHashSet<>();
     for (final TopicPartition p : partitions) {
       final OptionalLong start = startOffsets.apply(p);
       if (start.isPresent()) {
-        delegate.seek(p, start.getAsLong());
+        super.seek(p, start.getAsLong());
       } else {
         notFound.add(p);
       }
     }
-    return Collections.unmodifiableList(notFound);
+    uninitializedOffsets.removeAll(partitions);
+    return Collections.unmodifiableSet(notFound);
   }
 
   @Override
   public void assign(Collection<TopicPartition> partitions) {
+    for (final TopicPartition changelog : partitions) {
+      if (!super.assignment().contains(changelog)) {
+        uninitializedOffsets.add(changelog);
+      }
+    }
     super.assign(partitions);
-    seekInitial(partitions);
```

Comment:
This is the main fix -- keep `assign` as a lightweight metadata operation and just track any new partitions we received that may not yet have been initialized, then remove them from this set as we seek to the starting offset for each changelog partition.

During `poll` this gets checked and an exception is thrown if we find any uninitialized changelogs, to validate the assumptions about how the restore consumer is used within Streams.

There are a lot of alternative ways we could handle this, @rodesai I know you had another suggestion, and I'm not married to this "track offset initialization" strategy -- just picked it in the end because I think it will go well with the fix for standby tasks, where we'll be tracking offset initialization anyway. Happy to consider other options though

Reply:
Added some unit tests -- was struggling a bit to find a way to test this end-to-end and could only come up with some overly complex schemes, are the unit tests good enough for now?

I might be able to write better tests for this once we have the restore metrics in place, which I'm working on now/after this

Reply:
> There are a lot of alternative ways we could handle this, @rodesai I know you had another suggestion, and I'm not married to this "track offset initialization" strategy -- just picked it in the end because I think it will go well with the fix for standby tasks, where we'll be tracking offset initialization anyway. Happy to consider other options though

I'm happy with it as long as we're checking the assumption